File: app/src/main/java/it/cammino/risuscito/Utility.java
package it.cammino.risuscito;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Environment;
import android.support.v4.content.ContextCompat;
import android.util.Log;

import java.io.File;
import java.text.DecimalFormat;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

public class Utility {

    //Costanti per le impostazioni
    public static final String SCREEN_ON = "sempre_acceso";
    public static final String SHOW_SECONDA = "mostra_seconda_lettura";
    public static final String SHOW_PACE = "mostra_canto_pace";
    public static final String SAVE_LOCATION = "memoria_salvataggio_scelta";
    public static final String DEFAULT_INDEX = "indice_predefinito";
    public static final String SHOW_SANTO = "mostra_santo";
    public static final String SHOW_AUDIO = "mostra_audio";
    public static final String SYSTEM_LANGUAGE = "lingua_sistema";
    public static final String DB_RESET = "db_reset";
    public static final String CHANGE_LANGUAGE = "changed";
    public static final String SIGNED_IN = "signed_id";
    public static final String SHOW_OFFERTORIO = "mostra_canto_offertorio";
    public static final String SMALL_DONATIONS_COUNT = "small_donations_count";
    public static final String MEDIUM_DONATIONS_COUNT = "medium_donations_count";
    public static final String LARGE_DONATIONS_COUNT = "large_donations_count";

    public static final String INTRO_CONSEGNATI = "intro_consegnati_test";
    public static final String INTRO_CONSEGNATI_2 = "intro_consegnati_test";
    public static final String INTRO_PAGINARENDER = "intro_paginarender_test";
    public static final String INTRO_CREALISTA = "intro_crealista_test";
    public static final String INTRO_CREALISTA_2 = "intro_crealista_2_test";
    public static final String INTRO_CUSTOMLISTS = "intro_customlists_test_2";


//    public static final String GENERATE_XML = "generate_xml";

//    public static final int HIDE_DELAY = 1500;

    public static final long CLICK_DELAY = 1000;

    //Costanti per il passaggio dati alla pagina di visualizzazione canto in fullscreen
    public static final String URL_CANTO = "urlCanto";
    public static final String SPEED_VALUE = "speedValue";
    public static final String SCROLL_PLAYING = "scrollPlaying";
    public static final String ID_CANTO = "idCanto";
//    public static final String TAG_TRANSIZIONE = "fullscreen";
    public static final String TRANS_PAGINA_RENDER = "paginarender";

//    public static final String GIALLO = "#EBD0A5";
//    public static final String BIANCO = "#FCFCFC";
//    public static final String AZZURRO = "#6F949A";
//    public static final String VERDE = "#8FC490";
//    public static final String GRIGIO = "#CAC8BC";

    public static final int EXTERNAL_FILE_RC = 122;
    public static final int WRITE_STORAGE_RC = 123;
//    public static final int PHONE_LISTENER_RC = 124;

//    public static final String AUDIO_REQUESTED = "AUDIO_REQUESTED";

    //metodo che restituisce la stringa di input senza la pagina all'inizio
//    public static String truncatePage(String input) {
//
//        int length = input.length();
//        int start;
//
//        for (start = 0; start < length; start++) {
//
//            if (input.charAt(start) == ')') {
//                start += 2;
//                break;
//            }
//        }
//
//        return input.substring(start);
//    }

    //metodo che duplica tutti gli apici presenti nella stringa
    public static String duplicaApostrofi(String input) {

        String result = input;
        int massimo  = result.length() - 1;
        char apice = '\'';

        for (int i = 0; i <= massimo; i++) {
            if (result.charAt(i) == apice ) {
                result = result.substring(0, i+1) + apice + result.substring(i+1);
                massimo++;
                i++;
            }
        }

        return result;
    }

    public static String intToString(int num, int digits) {
//        assert digits > 0 : "Invalid number of digits";
        if (BuildConfig.DEBUG && !(digits > 0))
            throw new AssertionError("Campo digits non valido");

        // create variable length array of zeros
        char[] zeros = new char[digits];
        Arrays.fill(zeros, '0');
        // format number as String
        DecimalFormat df = new DecimalFormat(String.valueOf(zeros));

        return df.format(num);
    }

    public static boolean isOnline(Activity activity) {
        ConnectivityManager cm =
                (ConnectivityManager) activity.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = cm.getActiveNetworkInfo();
        return netInfo != null && netInfo.isConnectedOrConnecting();
    }

    /* Checks if external storage is available for read and write */
    public static boolean isExternalStorageWritable() {
        String state = Environment.getExternalStorageState();
        return Environment.MEDIA_MOUNTED.equals(state);
    }

    /* Checks if external storage is available to at least read */
    public static boolean isExternalStorageReadable() {
        String state = Environment.getExternalStorageState();
        return Environment.MEDIA_MOUNTED.equals(state) ||
                Environment.MEDIA_MOUNTED_READ_ONLY.equals(state);
    }

    /* Filtra il link di input per tenere solo il nome del file */
    public static String filterMediaLinkNew(String link) {
        if (link.length() == 0)
            return link;
        else {
            if (link.indexOf(".com") > 0) {
                int start = link.indexOf(".com/");
                return link.substring(start + 5).replaceAll("%20", "_");
            }
            else
            if (link.indexOf("ITALIANO/") > 0) {
                int start = link.indexOf("ITALIANO/");
                return link.substring(start + 9).replaceAll("%20", "_");
            }
            else
                return link;
        }
    }

    /* Filtra il link di input per tenere solo il nome del file */
    public static String filterMediaLink(String link) {
        if (link.length() == 0)
            return link;
        else {
            if (link.indexOf(".com") > 0) {
                int start = link.indexOf(".com/");
                return link.substring(start + 5);
            }
            else
            if (link.indexOf("ITALIANO/") > 0) {
                int start = link.indexOf("ITALIANO/");
                return link.substring(start + 9);
            }
            else
                return link;
        }
    }

    public static String retrieveMediaFileLink(Context activity, String link, boolean cercaEsterno) {

        if (isExternalStorageReadable() && cercaEsterno) {
//			File[] fileArray = ContextCompat.getExternalFilesDirs(activity, null);
//			File fileExt = new File(fileArray[0], filterMediaLink(link));
            //cerca file esterno con nuovi path e nome
            File fileExt = new File(Environment.getExternalStoragePublicDirectory(
                    Environment.DIRECTORY_MUSIC), "/Risuscitò/" + filterMediaLinkNew(link));
            if (fileExt.exists()) {
                Log.d("Utility.java", "FILE esterno: " + fileExt.getAbsolutePath());
                return fileExt.getAbsolutePath();
            } else {
                //cerca file esterno con vecchi path e nome
                File[] fileArray = ContextCompat.getExternalFilesDirs(activity, null);
                fileExt = new File(fileArray[0], filterMediaLink(link));
                if (fileExt.exists()) {
                    Log.d("Utility.java", "FILE esterno: " + fileExt.getAbsolutePath());
                    return fileExt.getAbsolutePath();
                }
                else
                    Log.d("Utility.java", "FILE ESTERNO NON TROVATO");
            }
        }
        else {
            Log.d("Utility.java", "isExternalStorageReadable: FALSE");
        }

        File fileInt = new File(activity.getFilesDir(), filterMediaLink(link));
        if (fileInt.exists()) {
			Log.d("Utility.java", "FILE interno: " + fileInt.getAbsolutePath());
            return fileInt.getAbsolutePath();
        }
        else
            Log.d("Utility.java", "FILE INTERNO NON TROVATO");
//		Log.i("FILE INTERNO:", "NON TROVATO");
        return "";
    }

    @SuppressLint("NewApi")
    public static void setupTransparentTints(Activity context, int color, boolean hasNavDrawer) {

//        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.KITKAT) {
//            SystemBarTintManager tintManager = new SystemBarTintManager(context);
//            tintManager.setStatusBarTintEnabled(true);
//            tintManager.setStatusBarTintColor(color);
//        }

        if (!hasNavDrawer && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
            context.getWindow().setStatusBarColor(color);
    }

//    @SuppressWarnings("ResourceType")
//    public static void blockOrientation(Activity activity) {
//        // Copied from Android docs, since we don't have these values in Froyo 2.2
//        int SCREEN_ORIENTATION_REVERSE_LANDSCAPE = 8;
//        int SCREEN_ORIENTATION_REVERSE_PORTRAIT = 9;
//
//        Display display = ((WindowManager) activity.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
//        int rotation = display.getRotation();
//        switch(activity.getResources().getConfiguration().orientation)
//        {
//            case Configuration.ORIENTATION_LANDSCAPE:
//                if(rotation == Surface.ROTATION_0 || rotation == Surface.ROTATION_90)
//                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
//                else
//                    activity.setRequestedOrientation(SCREEN_ORIENTATION_REVERSE_LANDSCAPE);
//                break;
//            case Configuration.ORIENTATION_PORTRAIT:
//                if(rotation == Surface.ROTATION_0 || rotation == Surface.ROTATION_270)
//                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
//                else
//                    activity.setRequestedOrientation(SCREEN_ORIENTATION_REVERSE_PORTRAIT);
//        }
//    }

    public static int random(int start, int end) {
        return ((new Random()).nextInt(end - start + 1) + start);
    }

    public static boolean hasMarshmallow() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
    }

    public static boolean isLowerCase(char ch) {
        return ch >= 'a' && ch <= 'z';
    }

//    public static boolean isUpperCase(char ch) {
//        return ch >= 'A' && ch <= 'Z';
//    }

    private static Map<Character, Character> MAP_NORM;

    public static String removeAccents(String value) {

        if (MAP_NORM == null || MAP_NORM.size() == 0)
        {
            MAP_NORM = new HashMap<>();
            MAP_NORM.put('\u00C1', 'A');
            MAP_NORM.put('\u00C0', 'A');
            MAP_NORM.put('\u00C2', 'A');
            MAP_NORM.put('\u00C3', 'A');
            MAP_NORM.put('\u00C4', 'A');
            MAP_NORM.put('\u00C8', 'E');
            MAP_NORM.put('\u00C9', 'E');
            MAP_NORM.put('\u00CA', 'E');
            MAP_NORM.put('\u00CB', 'E');
            MAP_NORM.put('\u00CD', 'I');
            MAP_NORM.put('\u00CC', 'I');
            MAP_NORM.put('\u00CE', 'I');
            MAP_NORM.put('\u00CF', 'I');
            MAP_NORM.put('\u00D9', 'U');
            MAP_NORM.put('\u00DA', 'U');
            MAP_NORM.put('\u00DB', 'U');
            MAP_NORM.put('\u00DC', 'U');
            MAP_NORM.put('\u00D2', 'O');
            MAP_NORM.put('\u00D3', 'O');
            MAP_NORM.put('\u00D4', 'O');
            MAP_NORM.put('\u00D5', 'O');
            MAP_NORM.put('\u00D6', 'O');
            MAP_NORM.put('\u00D1', 'N');
            MAP_NORM.put('\u00C7', 'C');
            MAP_NORM.put('\u00AA', 'A');
            MAP_NORM.put('\u00BA', 'O');
            MAP_NORM.put('\u00A7', 'S');
            MAP_NORM.put('\u00B3', '3');
            MAP_NORM.put('\u00B2', '2');
            MAP_NORM.put('\u00B9', '1');
            MAP_NORM.put('\u00E0', 'a');
            MAP_NORM.put('\u00E1', 'a');
            MAP_NORM.put('\u00E2', 'a');
            MAP_NORM.put('\u00E3', 'a');
            MAP_NORM.put('\u00E4', 'a');
            MAP_NORM.put('\u00E8', 'e');
            MAP_NORM.put('\u00E9', 'e');
            MAP_NORM.put('\u00EA', 'e');
            MAP_NORM.put('\u00EB', 'e');
            MAP_NORM.put('\u00ED', 'i');
            MAP_NORM.put('\u00EC', 'i');
            MAP_NORM.put('\u00EE', 'i');
            MAP_NORM.put('\u00EF', 'i');
            MAP_NORM.put('\u00F9', 'u');
            MAP_NORM.put('\u00FA', 'u');
            MAP_NORM.put('\u00FB', 'u');
            MAP_NORM.put('\u00FC', 'u');
            MAP_NORM.put('\u00F2', 'o');
            MAP_NORM.put('\u00F3', 'o');
            MAP_NORM.put('\u00F4', 'o');
            MAP_NORM.put('\u00F5', 'o');
            MAP_NORM.put('\u00F6', 'o');
            MAP_NORM.put('\u00F1', 'n');
            MAP_NORM.put('\u00E7', 'c');
        }

        if (value == null) {
            return "";
        }

        StringBuilder sb = new StringBuilder(value);

        for(int i = 0; i < value.length(); i++) {
            Character c = MAP_NORM.get(sb.charAt(i));
            if(c != null)
                sb.setCharAt(i, c);
        }

        return sb.toString();
    }

//    public static String escapeEmail(String email) {
//        String result = email.substring(0, email.indexOf("@"));
////        Log.d("UTILITY", "escapeEmail: email.indexOf @ " + email.indexOf("@"));
////        Log.d("UTILITY", "escapeEmail: result " + result);
//        return result.replaceAll("\\.", "_");
//    }

}

File: app/src/main/java/it/cammino/risuscito/Utility.kt
package it.cammino.risuscito

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.app.Activity
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.media.MediaScannerConnection
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.view.View
import android.view.WindowInsetsController
import android.view.WindowManager
import androidx.annotation.ColorInt
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import androidx.preference.PreferenceManager
import com.mikepenz.fastadapter.ui.utils.StringHolder
import com.mikepenz.materialdrawer.holder.ColorHolder
import it.cammino.risuscito.LUtils.Companion.hasQ
import it.cammino.risuscito.utils.ThemeUtils
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.text.Normalizer
import java.util.*
import java.util.regex.Pattern

object Utility {

    // Costanti per le impostazioni
    private val TAG = Utility::class.java.canonicalName
    const val SCREEN_ON = "sempre_acceso"
    const val SYSTEM_LANGUAGE = "lingua_sistema_new"
    const val DB_RESET = "db_reset"
    const val CHANGE_LANGUAGE = "changed"
    const val CLICK_DELAY: Long = 500
    internal const val SHOW_SECONDA = "mostra_seconda_lettura"
    internal const val SHOW_PACE = "mostra_canto_pace"
    internal const val SAVE_LOCATION = "memoria_salvataggio_scelta"
    internal const val DEFAULT_INDEX = "indice_predefinito"
    internal const val DEFAULT_SEARCH = "ricerca_predefinita"
    internal const val SHOW_SANTO = "mostra_santo"
    internal const val SHOW_AUDIO = "mostra_audio"
    internal const val SIGNED_IN = "signed_id"
    internal const val SIGN_IN_REQUESTED = "sign_id_requested"
    internal const val SHOW_OFFERTORIO = "mostra_canto_offertorio"
    internal const val PREFERITI_OPEN = "preferiti_open"
    internal const val HISTORY_OPEN = "history_open"
    internal const val INTRO_CONSEGNATI = "intro_consegnati_test"
    internal const val INTRO_CONSEGNATI_2 = "intro_consegnati_2_test"
    internal const val INTRO_PAGINARENDER = "intro_paginarender_test"
    internal const val INTRO_CREALISTA = "intro_crealista_test"
    internal const val INTRO_CREALISTA_2 = "intro_crealista_2_test"
    internal const val INTRO_CUSTOMLISTS = "intro_customlists_test_2"
    internal const val NIGHT_MODE = "night_mode"
    internal const val PRIMARY_COLOR = "new_primary_color"
    internal const val SECONDARY_COLOR = "new_accent_color"
    internal const val ULTIMA_APP_USATA = "ULTIMA_APP_USATA"
    internal const val CLICK_DELAY_SELECTION: Long = 300

    // Costanti per il passaggio dati alla pagina di visualizzazione canto in fullscreen
    internal const val HTML_CONTENT = "htmlContent"
    internal const val SPEED_VALUE = "speedValue"
    internal const val SCROLL_PLAYING = "scrollPlaying"
    internal const val ID_CANTO = "idCanto"
    internal const val PAGINA = "pagina"
    internal const val TIPO_LISTA = "tipoLista"
    internal const val WRITE_STORAGE_RC = 123


    /* Checks if external storage is available for read and write */
    internal val isExternalStorageWritable: Boolean
        get() {
            val state = Environment.getExternalStorageState()
            return Environment.MEDIA_MOUNTED == state
        }

    /* Checks if external storage is available to at least read */
    internal val isExternalStorageReadable: Boolean
        get() {
            val state = Environment.getExternalStorageState()
            return Environment.MEDIA_MOUNTED == state || Environment.MEDIA_MOUNTED_READ_ONLY == state
        }

    internal fun isDefaultLocationPublic(context: Context): Boolean {
        return Integer.parseInt(PreferenceManager.getDefaultSharedPreferences(context).getString(SAVE_LOCATION, "0")
                ?: "0") == 1
    }

    internal fun isOnline(context: Context): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        return if (LUtils.hasM())
            isOnlineM(connectivityManager)
        else isOnlineLegacy(connectivityManager)
    }

    @TargetApi(Build.VERSION_CODES.M)
    private fun isOnlineM(connectivityManager: ConnectivityManager?): Boolean {
        val network = connectivityManager?.activeNetwork
        val capabilities = connectivityManager?.getNetworkCapabilities(network)
        return capabilities?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
    }

    @Suppress("DEPRECATION")
    private fun isOnlineLegacy(connectivityManager: ConnectivityManager?): Boolean {
        return connectivityManager?.activeNetworkInfo?.isConnected == true
    }

    /* Filtra il link di input per tenere solo il nome del file */
    private fun filterMediaLinkNew(link: String?): String {
        link?.let {
            return if (it.isEmpty())
                it
            else {
                return when {
                    it.indexOf("resuscicanti") > 0 -> {
                        val start = it.indexOf(".com/")
                        it.substring(start + 5).replace("%20".toRegex(), "_")
                    }
                    it.indexOf("marbat87") > 0 -> {
                        val start = it.indexOf("audio/")
                        it.substring(start + 6).replace("%20".toRegex(), "_")
                    }
                    else -> it
                }
            }
        } ?: return ""
    }

    /* Filtra il link di input per tenere solo il nome del file */
    internal fun filterMediaLink(link: String?): String {
        link?.let {
            return if (it.isEmpty())
                it
            else {
                when {
                    it.indexOf("resuscicanti") > 0 -> {
                        val start = it.indexOf(".com/")
                        it.substring(start + 5)
                    }
                    it.indexOf("marbat87") > 0 -> {
                        val start = it.indexOf("audio/")
                        it.substring(start + 6)
                    }
                    else -> it
                }
            }
        } ?: return ""
    }

    fun retrieveMediaFileLink(activity: Context, link: String?, cercaEsterno: Boolean): String {

        if (link.isNullOrEmpty()) return ""

        return if (hasQ())
            retrieveMediaFileLinkQ(activity, link, cercaEsterno)
        else
            retrieveMediaFileLinkLegacy(activity, link, cercaEsterno)
    }

    @TargetApi(Build.VERSION_CODES.Q)
    fun retrieveMediaFileLinkQ(activity: Context, link: String, cercaEsterno: Boolean): String {

        if (isExternalStorageReadable && cercaEsterno) {
//            Log.d(TAG, "retrieveMediaFileLinkQ: " + getExternalLink(link))
            val externalId = getExternalMediaIdByName(activity, link)
            if (externalId >= 0) {
//                Log.d(TAG, "retrieveMediaFileLinkQ: FILE ESTERNO TROVATO")
                return getExternalLink(link)
            }
//            else
//                Log.d(TAG, "retrieveMediaFileLinkQ: FILE ESTERNO NON TROVATO")
        }
//        else
//            Log.d(TAG, "retrieveMediaFileLinkQ isExternalStorageReadable: FALSE")

        return retrieveInternalLink(activity, link)
    }

    @TargetApi(Build.VERSION_CODES.Q)
    internal fun getExternalMediaIdByName(context: Context, link: String): Long {
        val projection = arrayOf(MediaStore.Audio.Media.DISPLAY_NAME, MediaStore.Audio.Media._ID)
        val resolver = context.contentResolver
        val collection = MediaStore.Audio.Media
                .getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
        resolver.query(collection, projection, "${MediaStore.Audio.Media.DISPLAY_NAME} = ?", arrayOf(getExternalLink(link)), null).use { cursor ->
            cursor?.let {
                if (it.moveToFirst()) {
                    val nameColumn = it.getColumnIndexOrThrow(MediaStore.Audio.Media.DISPLAY_NAME)
                    Log.d(TAG, "retrieveMediaFileLinkQ DISPLAY_NAME: ${it.getString(nameColumn)}")
                    val idColumn = it.getColumnIndexOrThrow(MediaStore.Audio.Media._ID)
                    Log.d(TAG, "retrieveMediaFileLinkQ _ID: ${it.getInt(idColumn)}")
                    return it.getLong(idColumn)
                }
            }
        }
        return -1
    }

    @Suppress("DEPRECATION")
    fun retrieveMediaFileLinkLegacy(activity: Context, link: String, cercaEsterno: Boolean): String {

        if (isExternalStorageReadable && cercaEsterno) {
//            Log.d(TAG, "retrieveMediaFileLinkLegacy: " + filterMediaLinkNew(link))
            // cerca file esterno con nuovi path e nome
            var fileExt = File(
                    Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC),
                    "/Risuscitò/" + filterMediaLinkNew(link))
            if (fileExt.exists()) {
//                Log.d(TAG, "retrieveMediaFileLinkLegacy FILE esterno1: " + fileExt.absolutePath)
                return fileExt.absolutePath
            } else {
                // cerca file esterno con vecchi path e nome
                val fileArray = ContextCompat.getExternalFilesDirs(activity, null)
                fileExt = File(fileArray[0], filterMediaLink(link))
                if (fileExt.exists()) {
//                    Log.d(TAG, "retrieveMediaFileLinkLegacy FILE esterno2: " + fileExt.absolutePath)
                    return fileExt.absolutePath
                }
//                else
//                    Log.d(TAG, "retrieveMediaFileLinkLegacy FILE ESTERNO NON TROVATO")
            }
        }
//        else
//            Log.d(TAG, "retrieveMediaFileLinkLegacy isExternalStorageReadable: FALSE")

        return retrieveInternalLink(activity, link)
    }

    private fun retrieveInternalLink(activity: Context, link: String?): String {
        val fileInt = File(activity.filesDir, filterMediaLink(link))
        if (fileInt.exists()) {
//            Log.d(TAG, "FILE interno: " + fileInt.absolutePath)
            return fileInt.absolutePath
        }
//        else
//            Log.v(TAG, "FILE INTERNO NON TROVATO")
        //		Log.i("FILE INTERNO:", "NON TROVATO");
        return ""
    }

    @SuppressLint("NewApi")
    fun setupTransparentTints(context: Activity, color: Int, hasNavDrawer: Boolean, isOnTablet: Boolean) {
        if ((!hasNavDrawer || isOnTablet) && LUtils.hasL())
            context.window.statusBarColor = color
    }

    @SuppressLint("NewApi")
    fun setupNavBarColor(context: Activity) {
        context.window.decorView.setBackgroundColor(ContextCompat.getColor(context, if (ThemeUtils.isDarkMode(context)) R.color.design_dark_default_color_background else R.color.design_default_color_background))
        if (LUtils.hasO()) {
//            context.window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS and WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            context.window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            if (!ThemeUtils.isDarkMode(context)) setLightNavigationBar(context)
//            context.window.decorView.setBackgroundColor(ContextCompat.getColor(context, if (ThemeUtils.isDarkMode(context)) R.color.design_dark_default_color_background else R.color.design_default_color_background))
            context.window.navigationBarColor = Color.TRANSPARENT
        }
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun setLightNavigationBar(context: Activity) {
        if (LUtils.hasR())
            setLightNavigationBarR(context)
        else
            setLightNavigationBarLegacy(context)
    }

    @Suppress("DEPRECATION")
    @RequiresApi(Build.VERSION_CODES.O)
    fun setLightNavigationBarLegacy(context: Activity) {
        context.window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
    }

    @RequiresApi(Build.VERSION_CODES.R)
    fun setLightNavigationBarR(context: Activity) {
        context.window.insetsController?.setSystemBarsAppearance(WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS, WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS)
    }

    internal fun random(start: Int, end: Int): Int {
        return Random().nextInt(end - start + 1) + start
    }

    internal fun isLowerCase(ch: Char): Boolean {
        return ch in 'a'..'z'
    }

    internal fun removeAccents(value: String): String {

        var normalized = value

        for ((mapKey, mapValue) in STROKE_LETTERS)
            normalized = normalized.replace(mapKey, mapValue)

        normalized = Normalizer.normalize(normalized, Normalizer.Form.NFD)
        val pattern = Pattern.compile("\\p{InCombiningDiacriticalMarks}+")
        return pattern.matcher(normalized).replaceAll("")
    }

    fun createNotificationChannelWrapper(applicationContext: Context, channelId: String, name: String, description: String) {
        if (LUtils.hasO()) createNotificationChannel(applicationContext, channelId, name, description)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createNotificationChannel(applicationContext: Context, channelId: String, name: String, description: String) {
        val mNotificationManager = applicationContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val mChannel = NotificationChannel(channelId, name, NotificationManager.IMPORTANCE_LOW)
        // Configure the notification channel.
        mChannel.description = description
        mChannel.setShowBadge(false)
        mChannel.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
        mNotificationManager.createNotificationChannel(mChannel)
    }

    fun <T> helperSetString(t: T): StringHolder = when (t) {
        is String -> StringHolder(t)
        is Int -> StringHolder(t)
        else -> throw IllegalArgumentException()
    }

    fun <T> helperSetColor(t: T): ColorHolder = when (t) {
        is String -> ColorHolder.fromColor(Color.parseColor(t))
        is @ColorInt Int -> ColorHolder.fromColor(t)
        else -> throw IllegalArgumentException()
    }

    fun getExternalLink(link: String): String {
        return if (hasQ())
            getExternalLinkQ(link)
        else
            getExternalLinkLegacy(link)
    }

    private fun getExternalLinkQ(link: String): String {
        return filterMediaLinkNew(link)
    }

    @Suppress("DEPRECATION")
    private fun getExternalLinkLegacy(link: String): String {
        if (File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC),
                        "Risuscitò")
                        .mkdirs())
            Log.d(TAG, "CARTELLA RISUSCITO CREATA")
        else
            Log.d(TAG, "CARTELLA RISUSCITO ESISTENTE")
        return (Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC)
                .absolutePath
                + "/Risuscitò/"
                + filterMediaLinkNew(link))
    }

    @Suppress("DEPRECATION")
    fun mediaScan(context: Context, link: String) {
        MediaScannerConnection.scanFile(
                context,
                arrayOf(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC)
                        .absolutePath
                        + "/Risuscitò/"
                        + filterMediaLinkNew(link)), null, null)
    }

    private fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        // Raw height and width of image
        val (height: Int, width: Int) = options.run { outHeight to outWidth }
        var inSampleSize = 1

        if (height > reqHeight || width > reqWidth) {

            val halfHeight: Int = height / 2
            val halfWidth: Int = width / 2

            // Calculate the largest inSampleSize value that is a power of 2 and keeps both
            // height and width larger than the requested height and width.
            while (halfHeight / inSampleSize >= reqHeight && halfWidth / inSampleSize >= reqWidth) {
                inSampleSize *= 2
            }
        }

        return inSampleSize
    }

    internal fun decodeSampledBitmapFromResource(
            res: Resources,
            resId: Int,
            reqWidth: Int,
            reqHeight: Int
    ): Bitmap {
        // First decode with inJustDecodeBounds=true to check dimensions
        return BitmapFactory.Options().run {
            inJustDecodeBounds = true
            BitmapFactory.decodeResource(res, resId, this)

            // Calculate inSampleSize
            inSampleSize = calculateInSampleSize(this, reqWidth, reqHeight)

            // Decode bitmap with inSampleSize set
            inJustDecodeBounds = false

            BitmapFactory.decodeResource(res, resId, this)
        }
    }

    internal fun readTextFromResource(ctx: Context, resourceID: String): String {
        val inputStream = ctx.resources.openRawResource(LUtils.getResId(resourceID, R.raw::class.java))
        val br = BufferedReader(InputStreamReader(inputStream, ECONDING_UTF8))
        var line: String? = br.readLine()
        val cantoTrasportato = StringBuffer()

        while (line != null) {
//            Log.d(TAG, "line: $line")
            cantoTrasportato.append(line)
            cantoTrasportato.append("\n")
            line = br.readLine()
        }
        br.close()
        Log.d(TAG, "readTextFromResource cantoTrasportato: $cantoTrasportato")
        return cantoTrasportato.toString()
    }

    private const val ECONDING_UTF8 = "utf-8"

    private val STROKE_LETTERS: Map<String, String> = mapOf(
            Pair("Ł", "L"),
            Pair("ł", "l"))

}
